go/cmd/dolt/commands/cherry-pick.go

```diff
@@ -0,0 +1,466 @@
+	"github.com/dolthub/dolt/go/libraries/doltcore/env"
+	"github.com/dolthub/dolt/go/libraries/utils/argparser"
+)
+
+var cherryPickDocs = cli.CommandDocumentationContent{
+	ShortDesc: `Apply the changes introduced by an existing commit.`,
+	LongDesc: `
+Updates tables in the clean working set with changes introduced in cherry-picked commit and creates a new commit with applied changes.
+
+dolt cherry-pick {{.LessThan}}commit{{.GreaterThan}}
+   To apply changes from an existing {{.LessThan}}commit{{.GreaterThan}} to current HEAD, the current working tree must be clean (no modifications from the HEAD commit). 
+   By default, cherry-pick creates new commit with applied changes.`,
+	Synopsis: []string{
+		`{{.LessThan}}commit{{.GreaterThan}}`,
+	},
+}
+
+type CherryPickCmd struct{}
+
+// Name is returns the name of the Dolt cli command. This is what is used on the command line to invoke the command
+func (cmd CherryPickCmd) Name() string {
+	return "cherry-pick"
+}
+
+// Description returns a description of the command
+func (cmd CherryPickCmd) Description() string {
+	return "Apply the changes introduced by an existing commit from different branch."
+}
+
+// CreateMarkdown creates a markdown file containing the helptext for the command at the given path
+func (cmd CherryPickCmd) CreateMarkdown(wr io.Writer, commandStr string) error {
+	ap := cli.CreateCherryPickArgParser()
+	return CreateMarkdown(wr, cli.GetCommandDocumentation(commandStr, cherryPickDocs, ap))
+}
+
+func (cmd CherryPickCmd) ArgParser() *argparser.ArgParser {
+	return cli.CreateCherryPickArgParser()
+}
+
+// EventType returns the type of the event to log
+func (cmd CherryPickCmd) EventType() eventsapi.ClientEventType {
+	return eventsapi.ClientEventType_CHERRY_PICK
+}
+
+// Exec executes the command
+func (cmd CherryPickCmd) Exec(ctx context.Context, commandStr string, args []string, dEnv *env.DoltEnv) int {
+	ap := cmd.ArgParser()
+	help, usage := cli.HelpAndUsagePrinters(cli.GetCommandDocumentation(commandStr, dumpDocs, ap))
+	apr := cli.ParseArgsOrDie(ap, args, help)
+
+	// This command creates a commit, so we need user identity
+	if !cli.CheckUserNameAndEmail(dEnv) {
+		return 1
+	}
+
+	// TODO : support single commit cherry-pick only for now
+	if apr.NArg() == 0 {
+		usage()
+		return 1
+	} else if apr.NArg() > 1 {
+		return HandleVErrAndExitCode(errhand.BuildDError("multiple commits not supported yet.").SetPrintUsage().Build(), usage)
+	}
+
+	// check for clean working state
+	headRoot, err := dEnv.HeadRoot(ctx)
+	if err != nil {
+		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	}
+	workingRoot, err := dEnv.WorkingRoot(ctx)
+	if err != nil {
+		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	}
+	headHash, err := headRoot.HashOf()
+	if err != nil {
+		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	}
+	workingHash, err := workingRoot.HashOf()
```

Comment:
Seems like we also need to check `dEnv.StagedRoot(ctx)`? Would be good to have two separate bats tests – one for a working root change and one for a staged root change.